research/models/deidenitfier.py

```diff
@@ -0,0 +1,113 @@
+"""A module for PII detection and de-identification."""
+
+from typing import Dict, List, Optional
+
+from presidio_analyzer import AnalyzerEngine, Pattern, PatternRecognizer
+from presidio_anonymizer import AnonymizerEngine
+from presidio_anonymizer.entities import OperatorConfig
+
+
+class Deidentifier:
+    """Class for PII detection and de-identification using Presidio."""
+
+    def __init__(self):
+        """Initialize the Presidio Analyzer and Anonymizer engines."""
+        self.analyzer = AnalyzerEngine()
+        self.anonymizer = AnonymizerEngine()
+
+    def add_custom_recognizer(
+        self, entity_name: str, regex_pattern: str, score: float = 0.85
+    ):
+        """
+        Add a custom PII entity recognizer using a regular expression.
+
+        Args:
+            entity_name: The name for the new entity (e.g., "CUSTOM_ID").
+            regex_pattern: The regex pattern to detect the entity.
+            score: The confidence score for the detection (0.0 to 1.0).
+        """
+        if not (0.0 <= score <= 1.0):
+            raise ValueError('Score must be between 0.0 and 1.0.')
+
+        # Create a recognizer from the provided pattern
+        custom_recognizer = PatternRecognizer(
+            supported_entity=entity_name,
+            patterns=[
+                Pattern(name=entity_name, regex=regex_pattern, score=score)
+            ],
+        )
+
+        self.analyzer.registry.add_recognizer(custom_recognizer)
```

Comment:
First check whether a recognizer for that supported_entity is already in presidio memory registry; if it is, remove it first. Clearing the old recognizer prevents duplicate definitions and avoids ambiguous matches before you register the new one.
Something like:
```python
        if self.analyzer.registry.get_recognizer(entity_name):
            self.analyzer.registry.remove_recognizer(entity_name)

        self.analyzer.registry.add_recognizer(
            PatternRecognizer(
                supported_entity=entity_name,
                patterns=[Pattern(name=entity_name, regex=regex_pattern, score=score)],
            )
        )
```